src/policy/fees_util.cpp

```diff
@@ -65,3 +67,34 @@ TxAncestorsAndDescendants GetTxAncestorsAndDescendants(const std::vector<Removed
     }
     return visited_txs;
 }
+
+node::LinearizationResult LinearizeTransactions(const std::vector<RemovedMempoolTransactionInfo>& txs_removed_for_block)
+{
+    // Cache all the transactions for efficient lookup
```

Comment:
In  [fees]: add method that linearize transaction removed from mempool:

i) How big could this cache get?

Reply:
ii) It seems sub-optimal to have two `Linearize*` functions/methods... :S

Reply:
iii) If I understand correctly, this means that our last-six-block estimator is not actually estimating vs last 6 blocks, but rather on "last six blocks **where transactions were in our mempool**" (and got removed), which could be substantially different.

That said, it should be a lower value, I presume. And therefore more conservative (if less accurate)?

Reply:
> i) How big could this cache get?**

I will have to analyze this, but in the worst-case scenario, it will only save the following information for each block transaction in memory for easy lookup:

- Reference
- Fees
- Size
- Height and ID

Since this data is saved for a short amount of time, I think the tradeoff is worth it ?

---

> **ii) It seems sub-optimal to have two Linearize* functions/methods... :S**

I don't think it is, though; they both do two different things. `GetTxAncestorsAndDescendants` only calculates the transaction ancestors and descendants. Then, `LinearizeTransactions` calls `GetTxAncestorsAndDescendants`, appends some missing information, and calls the mini miner for linearization. I did it this way so that I can test `GetTxAncestorsAndDescendants` atomically.

But yeah, you are right. This will be on my to-do list to simplify  all this after the approach/concept ack of the PR.


> but rather on "last six blocks where transactions were in our mempool" (and got removed), which could be substantially different.


Not quite it actually, it's estimating with data of transactions in the last 6 blocks that were in our mempool before the block arrived and then got removed upon the arrival of the block for `BLOCK` reason.